tlv/tlv_test.go

```diff
@@ -0,0 +1,552 @@
+			0x00, 0x00, 0x01,
+		},
+		expErr: tlv.NewTypeForDecodingErr(new(nodeAmts), "nodeAmts", 48, 49),
+		skipN2: true,
+	},
+	{
+		name: "n1's node_id is not a valid point",
+		bytes: []byte{0x03, 0x31,
+			0x04, 0x3d, 0xa0, 0x92, 0xf6, 0x98, 0x0e, 0x58, 0xd2,
+			0xc0, 0x37, 0x17, 0x31, 0x80, 0xe9, 0xa4, 0x65, 0x47,
+			0x60, 0x26, 0xee, 0x50, 0xf9, 0x66, 0x95, 0x96, 0x3e,
+			0x8e, 0xfe, 0x43, 0x6f, 0x54, 0xeb, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x02,
+		},
+		expErr: errors.New("invalid magic in compressed pubkey string: 4"),
+		skipN2: true,
+	},
+	{
+		name: "greater than encoding length for n1's nodeAmts",
+		bytes: []byte{0x03, 0x32,
+			0x02, 0x3d, 0xa0, 0x92, 0xf6, 0x98, 0x0e, 0x58, 0xd2,
+			0xc0, 0x37, 0x17, 0x31, 0x80, 0xe9, 0xa4, 0x65, 0x47,
+			0x60, 0x26, 0xee, 0x50, 0xf9, 0x66, 0x95, 0x96, 0x3e,
+			0x8e, 0xfe, 0x43, 0x6f, 0x54, 0xeb, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+		},
+		expErr: tlv.NewTypeForDecodingErr(new(nodeAmts), "nodeAmts", 50, 49),
+		skipN2: true,
+	},
+	{
+		name:   "unknown required type or n1",
+		bytes:  []byte{0x00, 0x00},
+		expErr: tlv.ErrUnknownRequiredType(0x00),
+		skipN2: true,
+	},
+	{
+		name:   "less than encoding length for n1's cltvDelta",
+		bytes:  []byte{0xfd, 0x00, 0x0fe, 0x00},
+		expErr: tlv.NewTypeForDecodingErr(new(uint16), "uint16", 0, 2),
+		skipN2: true,
+	},
+	{
+		name:   "less than encoding length for n1's cltvDelta",
+		bytes:  []byte{0xfd, 0x00, 0xfe, 0x01, 0x01},
+		expErr: tlv.NewTypeForDecodingErr(new(uint16), "uint16", 1, 2),
+		skipN2: true,
+	},
+	{
+		name:   "greater than encoding length for n1's cltvDelta",
+		bytes:  []byte{0xfd, 0x00, 0xfe, 0x03, 0x01, 0x01, 0x01},
+		expErr: tlv.NewTypeForDecodingErr(new(uint16), "uint16", 3, 2),
+		skipN2: true,
+	},
+	{
+		name:   "unknown even field for n1's namespace",
+		bytes:  []byte{0x0a, 0x00},
+		expErr: tlv.ErrUnknownRequiredType(0x0a),
+		skipN2: true,
+	},
+	{
+		name: "valid records but invalid ordering",
+		bytes: []byte{0x02, 0x08,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x26, 0x01,
+			0x01, 0x2a,
+		},
+		expErr: tlv.ErrStreamNotCanonical,
+		skipN2: true,
+	},
+	{
+		name: "duplicate tlv type",
+		bytes: []byte{0x02, 0x08,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x31, 0x02,
+			0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x51,
+		},
+		expErr: tlv.ErrStreamNotCanonical,
+		skipN2: true,
+	},
+	{
+		name:   "duplicate ignored tlv type",
+		bytes:  []byte{0x1f, 0x00, 0x1f, 0x01, 0x2a},
+		expErr: tlv.ErrStreamNotCanonical,
+		skipN2: true,
+	},
+	{
+		name:   "type wraparound",
+		bytes:  []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00},
+		expErr: tlv.ErrStreamNotCanonical,
+	},
+}
+
+func TestTLVDecodingFailures(t *testing.T) {
+	for _, test := range tlvDecodingFailureTests {
+		t.Run(test.name, func(t *testing.T) {
```

Comment:
The bolt spec has numbered the test cases. I was thinking, would it be worth to prefix `test.name` with an index in this loop?

Reply:
the formatting in the spec is a little weird, indeed it is numbered, but each case is actually two lines so they won't line up unless we number them as `2*i+1`

Reply:
Yes, I saw the strange numbering. It was just an idea, nothing important.